python/lsst/cp/pipe/ptc.py

```diff
@@ -367,11 +434,269 @@ def runDataRef(self, dataRef, visitPairs):
+            # Calculate the full covariances as in Astier+19. The variances for the PTC will be
+            # cov[0,0].
+
+            tupleCovariancesWithTags = self.computeCovariancesAstier(dataRef, visitPairs, detector)
+            # use the np.recarray to obtain the CovFit objects
+            covFits, covFitsNoB = fitData(tupleCovariancesWithTags, maxMu=self.config.maxMeanSignal,
+                                          maxMuElectrons=self.config.maxMeanSignal,
+                                          r=self.config.maximumRangeCovariancesAstier)
+
+            if self.config.makePlots:
+                self.plotCovariancesAstier(dataRef, covFits, covFitsNoB, tupleCovariancesWithTags,
+                                           log=self.log)
+
+            datasetCovAstier = PhotonTransferCurveCovAstierDataset()
+            datasetCovAstier.covariancesTuple = tupleCovariancesWithTags
+            datasetCovAstier.covariancesFits = covFits
+            datasetCovAstier.covariancesFitsWithNoB = covFitsNoB
+
+            dataset = self.getOutputPtcDataCovAstier(covFits)
+
+            datasetCovAstier.gain = dataset.gain
+            datasetCovAstier.noise = dataset.noise
+            datasetCovAstier.meanSignal = dataset.signal
+            datasetCovAstier.var = dataset.var
+            datasetCovAstier.varModel = dataset.varModel
+            datasetCovAstier.meanSignalAdu = dataset.signalAdu
+            datasetCovAstier.varAdu = dataset.varAdu
+            datasetCovAstier.varModelAdu = dataset.varModelAdu
+            datasetCovAstier.a00 = dataset.a00
+            datasetCovAstier.b00 = dataset.b00
+
+            self.log.info(f"Writing Astier+19 covariances data")
+            dataRef.put(datasetCovAstier, datasetType="photonTransferCurveCovAstierDataset")
+
+        else:
+            # Calculate the PTC in the standard way (variances vs mean), and linearity by fitting mean_signal
+            # vs expTime.
+
+            dataset, lookupTableArray = self.computeStandardPtcAndNonLinearity(dataRef, visitPairs, detector)
+            if self.config.makePlots:
+                self.plot(dataRef, dataset, ptcFitType=self.config.ptcFitType)
+
+            # Save data, PTC fit, and NL fit dictionaries
+            self.log.info(f"Writing PTC and NL data to {dataRef.getUri(write=True)}")
+            dataRef.put(dataset, datasetType="photonTransferCurveDataset")
+
+            butler = dataRef.getButler()
+            self.log.info("Writing linearizers: \n "
+                          "lookup table (linear component of polynomial fit), \n "
+                          "polynomial (coefficients for a polynomial correction), \n "
+                          "and squared linearizer (quadratic coefficient from polynomial)")
+
+            detName = detector.getName()
+            now = datetime.datetime.utcnow()
+            calibDate = now.strftime("%Y-%m-%d")
+
+            for linType, dataType in [("LOOKUPTABLE", 'linearizeLut'),
```

Comment:
Linearizers cannot be generated from the Astier+19 measurement of the covariance?  I thought it was included in the "ptc pipeline" diagram.

Reply:
Yes. That document/diagram (pipeline_astier_shape_of_ptc.pdf in DM-20070 in Jira) was made by Pierre Astier, to try to explain his whole code and paper. It turns out his original code does, at least, 3 major things:
1) Corrects for non-linearity using photodiode data
2) Corrects for CTI in a way which I don’t remember right now
3) Does the actual covariances calculations and fits those to the full model in the paper.
I decided to have this ticket focus on 3), and leave 1) and 2) to other parts of the ISR chain (the linearity correction and the CTI correction).